forgerock-openbanking-uk-aspsp-common/src/main/java/com/forgerock/openbanking/common/services/openbanking/event/EventResponseUtilService.java

```diff
@@ -0,0 +1,128 @@
+
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+/**
+ * Util Service class to handler the Event Notification API responses thru all versions<br />
+ */
+@Service
+@Slf4j
+public class EventResponseUtilService {
+
+    private final ResourceLinkService resourceLinkService;
+    private OBVersion version;
+    private boolean shouldHaveMetaSection;
+
+    public EventResponseUtilService(ResourceLinkService resourceLinkService) {
+        this.resourceLinkService = resourceLinkService;
+        this.shouldHaveMetaSection = true;
+        this.version = null;
+    }
+
+    public OBVersion getVersion() {
+        return version;
+    }
+
+    public EventResponseUtilService setVersion(OBVersion version) {
+        this.version = version;
+        return this;
+    }
+
+    public boolean isShouldHaveMetaSection() {
+        return shouldHaveMetaSection;
+    }
+
+    public EventResponseUtilService setShouldHaveMetaSection(boolean shouldHaveMetaSection) {
+        this.shouldHaveMetaSection = shouldHaveMetaSection;
+        return this;
+    }
+
+    /**
+     * Provides the response collection object {@link OBCallbackUrlsResponse1} filtered by version instanced
+     * @param frCallbackUrls collection {@link FRCallbackUrl1}
+     * @return response collection object {@link OBCallbackUrlsResponse1}
+     */
+    public OBCallbackUrlsResponse1 packageResponse(final Collection<FRCallbackUrl1> frCallbackUrls) {
+        return new OBCallbackUrlsResponse1()
+                .data(new OBCallbackUrlsResponseData1()
+                        .callbackUrl(
+                                frCallbackUrls.stream()
+                                        .filter(it -> it.obCallbackUrl.getData().getVersion().equals(version.getCanonicalVersion()))
+                                        .map(this::toOBCallbackUrlResponseData1)
+                                        .collect(Collectors.toList())
+                        )
+                ).meta(checkConditionMetaAndVersion() ? new Meta() : null)
+                .links(
+                        (checkConditionMetaAndVersion() ? toSelfLink(frCallbackUrls.stream().findFirst().get()) : null)
+                );
+    }
+
+    /**
+     * Create and provide the response object {@link OBCallbackUrlResponse1} for the callback API<br />
+     * If the version doesn't exist the 'meta' section will be null or empty
+     * @param frCallbackUrl
+     *         {@link FRCallbackUrl1}
+     * @return the callback object response
+     */
+    public OBCallbackUrlResponse1 packageResponse(FRCallbackUrl1 frCallbackUrl) {
+        return new OBCallbackUrlResponse1()
+                .data(toOBCallbackUrlResponseData1(frCallbackUrl))
+                .meta(checkConditionMetaAndVersion() ? new Meta() : null)
+                .links(checkConditionMetaAndVersion() ? toSelfLink(frCallbackUrl) : null);
+    }
+
+    public boolean allowOperationByVersion(String versionToCompare) {
```

Comment:
It took me a while to understand what this is for exactly. It looks like it's to ensure a TPP isn't trying to get/update/delete a persisted callbackUrl entity via an older API. For example if something is stored against v3.1, then you can't try and delete it via v3.0 - is that right?

In the controllers you have:

```
if(eventResponseUtilService.allowOperationByVersion(byId.get().obCallbackUrl.getData().getVersion())){
                log.debug("Deleting callback url: {}", byId.get());
                callbackUrlsRepository.deleteById(callbackUrlId);
}
```

Could you not just have?
```
OBVersion version = OBVersion.fromString(byId.get().obCallbackUrl.getData().getVersion());
if(version.isBeforeVersion(v3_0)){
                log.debug("Deleting callback url: {}", byId.get());
                callbackUrlsRepository.deleteById(callbackUrlId);
}
```